### PR TITLE
[Release] Run release script to change 3.0.0.dev0 to 2.9.0

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Runs the following script from the release process to update the release numbers:

```
# Adapt the following to the correct version number for this release.
# This updates versions in Python and C++ files (This should, at a minimum, update python/ray/init.py, [build-docker.sh](http://build-docker.sh/), and src/ray/common/constants.h). (changes from python/ray/_private/utils.py, python/ray/tests/test_runtime_env.py, dashboard/client/package-lock.json should not be committed)
grep -l -R "3.0.0.dev0" * | grep -v "yaml" | grep -v "rst" | xargs -n 1 gsed -i 's/3.0.0.dev0/2.5.0/g'

# Note: python/ray/tests/test_runtime_env.py should NOT be changed!!
git checkout python/ray/tests/test_runtime_env.py

# This updates versions in java files
grep -l -R "2.0.0-SNAPSHOT" * | grep -v "yaml" | grep -v "rst" | xargs -n 1 gsed -i 's/2.0.0-SNAPSHOT/2.5.0/g'
```

I manually skipped the change to `python/ray/_private/utils.py` as indicated in the instruction.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/requirements_compiled.txt [hunk 1]
```diff
@@ -1844,7 +1844,7 @@ querystring-parser==1.2.4
     #   raydp
     #   tune-sklearn
     #   xgboost-ray
-raydp==1.7.0b20231020.dev0
+raydp==1.7.0b2022.9.0
```
Comment (architkulkarni) [inline python/requirements_compiled.txt:1847]: I'm not sure how the command `grep -l -R "3.0.0.dev0" * | grep -v "yaml" | grep -v "rst" | xargs -n 1 gsed -i 's/3.0.0.dev0/2.9.0/g'` causes this change, but it does.  I will revert this change.  Any thoughts on how to fix the command?

File: doc/source/ray-overview/pip_freeze_ray-py39-cpu.txt [hunk 2]
```diff
@@ -121,7 +121,7 @@ python-dotenv==1.0.0
 pytz==2022.7.1
 PyWavelets==1.4.1
 PyYAML==6.0.1
-ray @ file:///home/ray/ray-3.0.0.dev0-cp39-cp39-manylinux2014_x86_64.whl#sha256=85cf93d21b91ecef4cf6935194d03ca0c3f907d76267ecd5ad1c6fa6c768717b
+ray @ file:///home/ray/ray-2.9.0-cp39-cp39-manylinux2014_x86_64.whl#sha256=85cf93d21b91ecef4cf6935194d03ca0c3f907d76267ecd5ad1c6fa6c768717b
```
Comment (architkulkarni) [inline doc/source/ray-overview/pip_freeze_ray-py39-cpu.txt:124]: Do we need to revert this change, or wait until some later step in the process and then update the `sha` value?

**Comment by architkulkarni** [comment]
It looks like the release process instruction has not been updated.  I will use https://github.com/ray-project/ray/pull/41625 as a reference and update the release instruction.

File: doc/source/ray-overview/pip_freeze_ray-py39-cpu.txt (hunk 2 above)
Comment (aslonnie) [reply doc/source/ray-overview/pip_freeze_ray-py39-cpu.txt]: yes, please revert.

File: doc/source/tune/examples/pbt_visualization/pbt_visualization.ipynb [hunk 3]
```diff
@@ -290,15 +290,15 @@
...
+       "RayContext(dashboard_url='', python_version='3.8.13', ray_version='2.9.0', ray_commit='{{RAY_COMMIT_SHA}}', address_info={'node_ip_address': '127.0.0.1', 'raylet_ip_address': '127.0.0.1', 'redis_address': None, 'object_store_address': '/tmp/ray/session_2022-09-14_11-43-30_517456_2265/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2022-09-14_11-43-30_517456_2265/sockets/raylet', 'webui_url': '', 'session_dir': '/tmp/ray/session_2022-09-14_11-43-30_517456_2265', 'metrics_export_port': 63865, 'gcs_address': '127.0.0.1:65035', 'address': '127.0.0.1:65035', 'dashboard_agent_listen_port': 52365, 'node_id': '7655cd51cd62a60359c7b4de1d1589eed86fddc71de12a517caccc42'})"
```
Comment (aslonnie) [inline doc/source/tune/examples/pbt_visualization/pbt_visualization.ipynb]: revert these? I do not think they are really required.

but I am also not sure how this notebook is being used. shouldn't these notebook output be re-generated dynamically?

File: python/requirements_compiled.txt (hunk 1 above)
Comment (aslonnie) [reply python/requirements_compiled.txt]: this should be reverted.

the `grep` thing is terribly broke imo...

File: release/k8s_tests/run_gcs_ft_on_k8s.py [hunk 4]
```diff
@@ -25,7 +25,7 @@ class TestScenario(enum.Enum):
 
 if os.environ.get("RAY_IMAGE") is not None:
     ray_image = os.environ.get("RAY_IMAGE")
-elif ray.__version__ != "3.0.0.dev0":
+elif ray.__version__ != "2.9.0":
```
Comment (aslonnie) [inline release/k8s_tests/run_gcs_ft_on_k8s.py]: this should be reverted.

File: release/ray_release/tests/test_wheels.py [hunk 5]
```diff
@@ -49,20 +49,20 @@ def test_get_ray_wheels_url(remove_buildkite_env):
         repo_url="https://github.com/ray-project/ray.git",
         branch="master",
         commit="1234",
-        ray_version="3.0.0.dev0",
+        ray_version="2.9.0",
```
Comment (aslonnie) [inline release/ray_release/tests/test_wheels.py]: all changes in this file should be reverted.

all these are unit test cases of `ray_release`.

File: release/ray_release/wheels.py [hunk 6]
```diff
@@ -158,7 +158,7 @@ def get_ray_wheels_url_from_local_wheel(ray_wheels: str) -> Optional[str]:
 
     The uploaded object will have local user and current timestamp encoded
     in the upload key path, e.g.:
-    "ubuntu_2022_01_01_23:59:99/ray-3.0.0.dev0-cp37-cp37m-manylinux_x86_64.whl"
+    "ubuntu_2022_01_01_23:59:99/ray-2.9.0-cp37-cp37m-manylinux_x86_64.whl"
```
Comment (aslonnie) [inline release/ray_release/wheels.py:161]: also please revert these.

File: rllib/tests/backward_compat/checkpoints/v1.0/dqn_frozenlake_tf/policies/default_policy/rllib_checkpoint.json [hunk 7]
```diff
@@ -1 +1 @@
-{"type": "Policy", "checkpoint_version": "1.0", "ray_version": "3.0.0.dev0", "ray_commit": "5ae1638c8797518dc2177d7cf5d327cd9c32d706"}
\ No newline at end of file
+{"type": "Policy", "checkpoint_version": "1.0", "ray_version": "2.9.0", "ray_commit": "5ae1638c8797518dc2177d7cf5d327cd9c32d706"}
```
Comment (aslonnie) [inline rllib/tests/backward_compat/checkpoints/v1.0/dqn_frozenlake_tf/policies/default_policy/rllib_checkpoint.json:1]: all the rllit tests and checkpoints stuff should be reverted.

**Review by aslonnie — COMMENTED** [review]
okay.. so a lot of things should be reverted. I think the `grep` thing should just be killed, and refactored to a set of specific places.

here are the things that must change:
- the version strings in `pom.xml` and `pom_template.xml` 
- the `kRayVersion` in `constants.h`
- the `__version__` in `_version.py`
- `ci/ray_ci/utils.py` # though I think this can be refactored out.. @can-anyscale  ?

all other places are probably optional.

the one in `test_reporter.py` test output should be referencing the `__version__` in `_version.py`, rather than hard-coding the version output.

**Comment by architkulkarni** [comment]
@aslonnie thanks for the careful review.  I have updated the PR to match https://github.com/ray-project/ray/pull/40515, with the following exceptions:

New in 2.9.0
-> test_reporter.py.  You mentioned this should be fixed by using version in _version.py instead of hardcoding. This makes sense, I think we can just fix it on master and only fix it here if it breaks a test, but let me know if you think otherwise.

Changed from 2.8.0 to 2.9.0
-> the __init__.py update is now in _version.py instead.

**Pull request merged**